src/js/actions/tools.js

```diff
@@ -419,9 +431,131 @@ define(function (require, exports) {
      * @private
      * @type {function()}
      */
-    var _toolModalStateChangedHandler;
+    var _toolModalStateChangedHandler,
+        _vectorMaskHandler;
     
     /**
+     * Change the tool's vector mode
+     * This function will both dispatch a vector_state_changed event for the store
+     * and modify the current tool in order to set its properties to mask mode. 
+     * 
+     * @param {boolean} vectorState
+     * @return {Promise}
+     */
+    var changeVectorState = function (vectorState) {
+        var flux = this.flux,
+            toolStore = flux.store("tool"),
+            prevVectorMode = toolStore.getVectorMode(),
+            vectorMode = vectorState,
+            currentTool = toolStore.getCurrentTool(),
+            firstLaunch = false;
+
+        if (prevVectorMode === vectorMode || !currentTool.handleVectorMaskMode) {
+            return Promise.resolve();
+        }
+
+        var appStore = flux.store("application"),
+            currentDocument = appStore.getCurrentDocument(),
+            currentLayers = currentDocument.layers.selected,
+            currentLayer = currentLayers.first(),
+            initPromise,
+            dispatchPromise;
+
+        if (!currentLayer) {
+            return Promise.resolve();
+        }
+        var nonVectorModeLayer = (currentLayer.kind === currentLayer.layerKinds.BACKGROUND ||
+                currentLayer.kind === currentLayer.layerKinds.VECTOR);
+
+        if (vectorMode && nonVectorModeLayer) {
+            return Promise.resolve();
+        }
+        if (!currentLayer.vectorMaskEnabled) {
+            initPromise = descriptor.playObject(vectorMaskLib.createRevealAllMask());
+            var payload = {
+                documentID: currentDocument.id,
+                layerIDs: Immutable.List.of(currentLayer.id),
+                vectorMaskEnabled: true
+            },
+                dispatchAdd = this.dispatch(events.document.history.optimistic.ADD_VECTOR_MASK_TO_LAYER, payload),
+                dispatchToolMode = this.dispatch(events.tool.VECTOR_STATE_CHANGE, vectorMode);
+            dispatchPromise = Promise.join(dispatchAdd, dispatchToolMode);
+        } else {
+            initPromise = Promise.resolve();
+            dispatchPromise = this.dispatch(events.tool.VECTOR_STATE_CHANGE, vectorMode);
+        }
+
+        var toolMode = toolLib.toolModes.SHAPE;
+
+        if (vectorMode) {
+            toolMode = toolLib.toolModes.PATH;
+        }
+        
+        if (currentTool.id === "rectangle" || currentTool.id === "ellipse" || currentTool.id === "pen") {
+            var setObj = toolLib.setShapeToolMode(toolMode),
+                resetPromise = descriptor.batchPlayObjects([setObj]);
+
+            
+            if (!vectorMode && firstLaunch) {
+                var fillColor = [217, 217, 217],
+                    strokeColor = [157, 157, 157];
+
+                var defaultPromise = this.transfer(installShapeDefaults,
```

Comment:
i did something similar, i dont see the advantage of having 2 actions instead of 1 for installShapeDefaults, so i added the defaults to that function